### PR TITLE
Replace generic CLIErrors with more specific exceptions where possible 

### PR DESCRIPTION
**Pull request by c-ryan-k**
This replaces generic `raise CLIError` code with specific exceptions from `azure.cli.core.azclierror`
 
Please review the exception types here: 
https://github.com/Azure/azure-cli/blob/dev/src/azure-cli-core/azure/cli/core/azclierror.py 
and validate that we're using the correct / best match for each scenario.

More details on CLI error guidelines:
https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md

---
This project has adopted the [Microsoft Open Source Code of Conduct](https://opensource.microsoft.com/codeofconduct/). For more information see the [Code of Conduct FAQ](https://opensource.microsoft.com/codeofconduct/faq/) or contact [opencode@microsoft.com](mailto:opencode@microsoft.com) with any additional questions or comments.

Thank you for contributing to the IoT extension!

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [ ] If introducing new functionality or modified behavior, are they backed by unit and integration tests?
- [ ] In the same context as above are command names and their parameter definitions accurate? Do help docs have sufficient content?
- [ ] Have **all** unit **and** integration tests passed locally? i.e. `pytest <project root> -vv`
- [ ] Have static checks passed using the .pylintrc and .flake8 rules? Look at the CI scripts for example usage.
- [ ] Have you made an entry in HISTORY.rst which concisely explains your feature or change?

### CODE REVIEW

File: azext_iot/digitaltwins/providers/twin.py [hunk 1]
```diff
@@ -92,7 +92,9 @@ def update(self, twin_id, json_patch, etag=None):
         elif isinstance(json_patch, list):
             json_patch_collection.extend(json_patch)
         else:
-            raise CLIError(f"--json-patch content must be an object or array. Actual type was: {type(json_patch).__name__}")
+            raise InvalidArgumentValueError(
+                f"--json-patch content must be an object or array. Actual type was: {type(json_patch).__name__}"
```
Comment (vilit1) [inline azext_iot/digitaltwins/providers/twin.py:96]: does type(json_patch) not work?

File: azext_iot/iothub/providers/job.py [hunk 2]
```diff
@@ -128,37 +132,37 @@ def create(
...
+            raise RequiredArgumentMissingError(
                 "The query condition is required when job type is {} or {}. "
                 "Use query condition '*' if you need to run job on all devices.".format(
                     JobType.scheduleUpdateTwin.value, JobType.scheduleDeviceMethod.value
                 )
             )
 
         if poll_duration < 1:
-            raise CLIError("--poll-duration must be greater than 0!")
+            raise InvalidArgumentValueError("--poll-duration must be greater than 0!")
```
Comment (vilit1) [inline azext_iot/iothub/providers/job.py:143]: Why is there an ! and not a . ?

Comment (digimaun) [reply azext_iot/iothub/providers/job.py:143]: The string content wasn't changed in this PR.

File: azext_iot/digitaltwins/providers/twin.py (hunk 1 above)
Comment (digimaun) [reply azext_iot/digitaltwins/providers/twin.py:96]: This wasn't changed in the PR. type(json_patch) gives a string representation of the object class (usually object.__class__) where-as .__name__ gives just the class name, which is useful when you want a specific msg format.

File: azext_iot/iothub/providers/job.py (hunk 2 above)
Comment (vilit1) [reply azext_iot/iothub/providers/job.py:143]: Just a general why - I do not expect a string change here but it does feel weird to have an !

Comment (digimaun) [reply azext_iot/iothub/providers/job.py:143]: Probably some small micro-decision that happened over 2 years ago.

File: azext_iot/common/utility.py [hunk 3]
```diff
@@ -134,7 +138,7 @@ def process_json_arg(content, argument_name, preserve_order=False):
             )
 
         file_content_error = "from file: '{}' ".format(json_from_file)
-        raise CLIError(
+        raise CLIInternalError(
             "Failed to parse json {}for argument '{}' with exception:\n    {}".format(
```
Comment (vilit1) [inline azext_iot/common/utility.py:142]: nit: can this typo be fixed

File: azext_iot/common/utility.py [hunk 4]
```diff
@@ -351,6 +355,36 @@ def unpack_msrest_error(e):
...
+        op_status = getattr(e.response, 'status_code', None)
+        if op_status in [400, 409]:
+            raise BadRequestError(err)
+        if op_status == 401:
+            raise UnauthorizedError(err)
+        if op_status == 403:
+            raise ForbiddenError(err)
+        if op_status == 404:
+            raise ResourceNotFoundError(err)
+        if op_status in [500, 503, 504]:
```
Comment (vilit1) [inline azext_iot/common/utility.py:379]: Should this list include all 5xx values?

File: azext_iot/common/utility.py [hunk 5]
```diff
@@ -351,6 +355,36 @@ def unpack_msrest_error(e):
     return op_err
 
 
+def handle_service_exception(e):
+    from azure.cli.core.azclierror import (
```
Comment (vilit1) [inline azext_iot/common/utility.py:359]: Would be nice to include a link to the error guidance documentation, especially if the azure cli team decides to change the types.

File: azext_iot/digitaltwins/providers/model.py [hunk 6]
```diff
@@ -59,7 +59,7 @@ def __init__(self, cmd, name, rg=None):
 
     def add(self, models=None, from_directory=None):
         if not any([models, from_directory]):
-            raise CLIError("Provide either --models or --from-directory.")
+            raise RequiredArgumentMissingError("Provide either --models or --from-directory.")
```
Comment (vilit1) [inline azext_iot/digitaltwins/providers/model.py:62]: nit: can possibly use recommendations here

File: azext_iot/iothub/providers/job.py (hunk 2 above)
Comment (vilit1) [reply azext_iot/iothub/providers/job.py:143]: nit: can we change this so it is less awk

File: azext_iot/operations/dps.py [hunk 7]
```diff
@@ -156,7 +163,7 @@ def iot_dps_device_enrollment_create(
 
         if attestation_type == AttestationType.tpm.value:
             if not endorsement_key:
-                raise CLIError("Endorsement key is requried")
+                raise RequiredArgumentMissingError("Endorsement key is requried")
```
Comment (vilit1) [inline azext_iot/operations/dps.py:166]: need to include the parameter needed --endorsement-key

File: azext_iot/common/utility.py (hunk 3 above)
Comment (digimaun) [reply azext_iot/common/utility.py:142]: What is the typo? I'm not seeing it right now.

File: azext_iot/iothub/providers/job.py (hunk 2 above)
Comment (digimaun) [reply azext_iot/iothub/providers/job.py:143]: Feel free to change it to period if you desire.

File: azext_iot/operations/dps.py (hunk 7 above)
Comment (digimaun) [reply azext_iot/operations/dps.py:166]: Also typo 'requried'  ->  'required'

File: azext_iot/central/providers/destination_provider.py [hunk 8]
```diff
@@ -57,7 +56,7 @@ def add_destination(
         AdxDestinationV1_1_preview,
     ]:
         if destination_id in self._destinations:
-            raise CLIError("Destination already exists")
+            raise BadRequestError("Destination already exists")
```
Comment (digimaun) [inline azext_iot/central/providers/destination_provider.py:59]: Hmm in this case the request to the service is well formed, but the user destination Id already exists...a '409' like error seems more appropriate compared to a 400 like error.

File: azext_iot/central/providers/destination_provider.py [hunk 9]
```diff
@@ -97,7 +96,7 @@ def update_destination(
         )
 
         if not destination:
-            raise CLIError(
+            raise AzureResponseError(
                 "Failed to create destination with id: '{}'.".format(destination_id)
```
Comment (digimaun) [inline azext_iot/central/providers/destination_provider.py:100]: Unrelated text fix: "Failed to **update** destination..."

File: azext_iot/central/providers/destination_provider.py (hunk 8 above)
Comment (c-ryan-k) [reply azext_iot/central/providers/destination_provider.py:59]: There were a few of these where I wasn't sure. There isn't really a `409 Conflict` type of azclierror, but I think you're right in that `ClientRequestError` seems a bit more apt here than `BadRequestError` - because the request is formed correctly, there's just an error handling it.

I'll update any of the `Resource already exists` type of errors to raise a `ClientRequestError` instead

Comment (c-ryan-k) [reply azext_iot/central/providers/destination_provider.py:59]: Or rather (now that I'm reading this again), should we only update the situations where we're hitting an internal cache to use a `ClientRequestError` - and leave other instances (`device create`, `policy create`, etc.) as a `BadRequestError` if the resources already exist?

File: azext_iot/common/utility.py (hunk 4 above)
Comment (c-ryan-k) [reply azext_iot/common/utility.py:379]: The intent was to just include 5XX error codes in this group that were actual internal service errors (500/503/504*) as `AzureInternalError` - and leave the rest to generic `AzureResponseErrors` which is more of a fallback.

\* 504 / 502 is iffy (gateway timeout / bad gateway) because they don't necessarily directly indicate errors in the service, but it could be an error passed from the resource itself when the service endpoint is acting as a gateway.

I can add 501 (not implemented), 502 (bad gateway), and 505 (http version not supported) here to be more inclusive.

@digimaun  - any thoughts?

File: azext_iot/central/providers/destination_provider.py (hunk 8 above)
Comment (digimaun) [reply azext_iot/central/providers/destination_provider.py:59]: Yea the internal cache cases where the client does not send the request are apt for `ClientRequestError`. This occurs elsewhere, in say ....`device create`

image

File: azext_iot/central/providers/device_provider.py [hunk 10]
```diff
@@ -114,7 +119,7 @@ def create_device(
         )
 
         if not device:
-            raise CLIError("No device found with id: '{}'.".format(device_id))
+            raise ResourceNotFoundError("No device found with id: '{}'.".format(device_id))
```
Comment (digimaun) [inline azext_iot/central/providers/device_provider.py:122]: Hrm, in this case we are in the context of create device, a falsey result from the service component indicates a failure. Two comments here:
- The exception text should say something like "Failure to create device with id ....".
- Since we don't have the response exception here a `AzureResponseError` could be appropriate to represent a generic service error response.

File: azext_iot/common/utility.py (hunk 4 above)
Comment (c-ryan-k) [reply azext_iot/common/utility.py:379]: To summarize an offline discussion, based on the documentation in the [ `azclierror` class](https://github.com/Azure/azure-cli/blob/dev/src/azure-cli-core/azure/cli/core/azclierror.py), we're going to have all 500-599 errors raise an `AzureInternalError`, and we're also moving 409 errors from `BadRequestError` to `AzureResponseError`, as the documentation for `BadRequestError` explicitly states only 400 responses.

File: azext_iot/digitaltwins/providers/resource.py [hunk 11]
```diff
@@ -141,9 +141,6 @@ def get(self, name, resource_group_name, wait=False):
                 resource_name=name, resource_group_name=resource_group_name
             )
         except ErrorResponseException as e:
-            if wait:
```
Comment (digimaun) [inline azext_iot/digitaltwins/providers/resource.py:137]: Let's ensure wait operations are not impacted cc: @vilit1

**Comment by digimaun** [comment]
@yang-hai-feng, @lucadruda are either of you able to spare some cycles to review this PR to ensure we are aligned on Central exception changes? The changes are around changing generic CLIError exceptions to more semantically correct ones.

File: azext_iot/digitaltwins/providers/resource.py (hunk 11 above)
Comment (vilit1) [reply azext_iot/digitaltwins/providers/resource.py:137]: they are not - the wait commands will wait for the resource to have a terminating provisioning state as before.

File: azext_iot/common/utility.py (hunk 3 above)
Comment (digimaun) [reply azext_iot/common/utility.py:142]: The string format is not a typo...look at the inputs to the placeholder, it is either the value of `file_content_error` which ends in a space, or the empty string.

File: azext_iot/central/providers/device_provider.py [hunk 12]
```diff
@@ -114,7 +120,7 @@ def create_device(
         )
 
         if not device:
-            raise CLIError("No device found with id: '{}'.".format(device_id))
+            raise AzureResponseError("Failed to create device with id: '{}'.".format(device_id))
```
Comment (yang-hai-feng) [inline azext_iot/central/providers/device_provider.py:123]: Nice catch, thanks for the fix!

File: azext_iot/central/services/destination.py [hunk 13]
```diff
@@ -154,7 +153,7 @@ def list_destinations(
         )
 
         if "value" not in result:
-            raise CLIError("Value is not present in body: {}".format(result))
+            raise CLIInternalError("Value is not present in body: {}".format(result))
```
Comment (yang-hai-feng) [inline azext_iot/central/services/destination.py:156]: Should be AzureResponseError

File: azext_iot/central/services/export.py [hunk 14]
```diff
@@ -136,7 +135,7 @@ def list_exports(
         )
 
         if "value" not in result:
-            raise CLIError("Value is not present in body: {}".format(result))
+            raise CLIInternalError("Value is not present in body: {}".format(result))
```
Comment (yang-hai-feng) [inline azext_iot/central/services/export.py:138]: AzureResponseError?

File: azext_iot/central/services/job.py [hunk 15]
```diff
@@ -106,7 +105,7 @@ def _list_job(
         result = _utility.try_extract_result(response)
 
         if "value" not in result:
-            raise CLIError("Value is not present in body: {}".format(result))
+            raise CLIInternalError("Value is not present in body: {}".format(result))
```
Comment (yang-hai-feng) [inline azext_iot/central/services/job.py:108]: AzureResponseError?

File: azext_iot/central/services/organization.py [hunk 16]
```diff
@@ -135,7 +134,7 @@ def list_orgs(
         )
 
         if "value" not in result:
-            raise CLIError("Value is not present in body: {}".format(result))
+            raise CLIInternalError("Value is not present in body: {}".format(result))
```
Comment (yang-hai-feng) [inline azext_iot/central/services/organization.py:137]: AzureResponseError?

File: azext_iot/central/services/user.py [hunk 17]
```diff
@@ -247,7 +248,7 @@ def get_user_list(
         )
 
         if "value" not in result:
-            raise CLIError("Value is not present in body: {}".format(result))
+            raise CLIInternalError("Value is not present in body: {}".format(result))
```
Comment (yang-hai-feng) [inline azext_iot/central/services/user.py:251]: AzureResponseError

File: azext_iot/central/services/destination.py (hunk 13 above)
Comment (c-ryan-k) [reply azext_iot/central/services/destination.py:156]: Gotcha, will update these - I think I may have gotten some of them confused with the `Error parsing body` issues elsewhere. In this instance, we're expecting the service response to have `value` as a property, and if it doesn't we're assuming that implies a service error.

**Review by digimaun — APPROVED** [review]

**Review by yang-hai-feng — APPROVED** [review]

**Comment by digimaun** [comment]
Needs an entry in HISTORY.rst, `0.14.0` is not yet released.
